active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+                           "apiAction": "/listClusterAdminCredential",
+                           "query": "?a=1&b=2" // optional
+                           "body": {} // optional
+                        },
+                        {
+                           "type": "JSONPath",
+                           "path": "kubeconfig[0].value"
+                        }
+                     ]
+                  }
+               }
+            },
+            "template": {
+               "extensions": {
+                  "k8s": {
+                     "extension": "Kubernetes",
+                     "version": "1.0.0"
+                     // "config" is removed, it's expected to be provided by properties of the deployment
+                  }
+               }
+            }
+         }
+      }
+   }
+}
+```
+
+### Microsoft.Resources/deployments API changes
+
+This REP extends upon the API changes in REP 0003 by adding a configuration value to the extensions returned from
+a deployment GET. This configuration object will be the same as the configuration object described in the above ARM
+template design changes.
+
+```json
+{
+   // ... deployment contract
+   "extensions": [
+      {
+         "name": "Kubernetes",
+         "alias": "k8s",
+         "version": "1.0.0",
+         "deploymentId": "/subscriptions/.../resourceGroups/.../providers/Microsoft.Resources/deployments/...",
+         "config": {
+            "namespace": {
+               "type": "string",
+               "value": "default"
+            },
+            "kubeConfig": {
+               "type": "directive",
+               "evaluation": [
+                  {
+                     "type": "ArmApiCall",
+                     "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+                     "method": "GET",
+                     "apiVersion": "2024-02-01",
+                     "apiAction": "/listClusterAdminCredential",
+                     "query": "?a=1&b=2" // optional
+                     "body": { } // optional
+                  },
+                  {
+                     "type": "JSONPath",
+                     "path": "kubeconfig[0].value" // decompose?
+                  }
+               ]
+            }
+         }
+      }
+   ]
+}
+```
+
+With the deployments service returning a list of all extensions (across all nested deployments) with their configuration
+directives and returning all extension resources deployed in the template (across all nested deployments), the stacks 
+service can process each extension resource and use the "deploymentId" and "extensionAlias" properties as a composite 
+key to the data in the extensions array.
+
+## Drawbacks
+
+The biggest drawback of this approach is that extension configuration data cannot be sourced from deployment parameters.
+This design decision is driven solely to support the stacks user experience not needing to prompt for the parameter 
```

Comment:
Related to the above question - could there be any need to parameterize non-secure configuration? E.g. being able to parameterize the k8s namespace. Would we be blocking that with this approach?

Reply:
As it's spelled out as of now, yes. I outline that a little bit below and in unresolved questions. I haven't had a chance to look into how the configuration object schema metadata is loaded or what its source is yet. If it could be modified to outline if a property is a secret/secure value or non-secret, that could potentially be used on the engine side to enable providing a resolved configuration back through the GET response. Namespace could be sourced from a non-secure parameter, and the value used for the deployment parameter can be persisted inside the configuration object for consumption by stacks.

The input extensions property to the deployment would need to be able to accept language expressions that would be evaluated at template evaluation time. There would need to be a way to throw an error if a secure value is used at template validation time.

Reply:
Adapted the design to accept language expressions. It should have been there from the start.